src/Swarm/TUI/View.hs

```diff
@@ -167,31 +168,34 @@ chooseCursor s locs = case s ^. uiState . uiModal of
   Nothing -> showFirstCursor s locs
   Just _ -> Nothing
 
--- | Width cap for modal and error message windows
-maxModalWindowWidth :: Int
-maxModalWindowWidth = 500
-
 -- | Render the error dialog window with a given error message
 renderErrorDialog :: Text -> Widget Name
 renderErrorDialog err = renderDialog (dialog (Just "Error") Nothing (maxModalWindowWidth `min` requiredWidth)) errContent
  where
   errContent = txtWrapWith indent2 {preserveIndentation = True} err
   requiredWidth = 2 + maximum (textWidth <$> T.lines err)
 
--- | Render a fullscreen widget with some padding
-renderModal :: Modal -> Widget Name
-renderModal modal = renderDialog (dialog (Just modalTitle) Nothing maxModalWindowWidth) modalWidget
+-- | Draw the error dialog window, if it should be displayed right now.
+drawDialog :: UIState -> Widget Name
+drawDialog s = case s ^. uiModal of
+  Just (Modal _ d w) -> renderDialog d w
+  Nothing -> maybe emptyWidget renderErrorDialog (s ^. uiError)
+
+-- | Generate a fresh modal window of the requested type.
+generateModal :: ModalType -> Modal
+generateModal mt = Modal mt (dialog (Just title) buttons (maxModalWindowWidth `min` requiredWidth)) widget
  where
-  modalWidget = Widget Fixed Fixed $ do
-    ctx <- getContext
-    let w = ctx ^. availWidthL
-        h = ctx ^. availHeightL
-        padding = 10
-    render $ setAvailableSize (w - padding, h - padding) modalContent
-  (modalTitle, modalContent) =
-    case modal of
-      HelpModal -> ("Help", helpWidget)
-      WinModal -> ("", txt "Congratulations!")
+  (title, widget, buttons, requiredWidth) =
+    case mt of
+      HelpModal -> ("Help", helpWidget, Nothing, maxModalWindowWidth)
+      WinModal -> ("", txt "Congratulations!", Nothing, maxModalWindowWidth)
+      QuitModal ->
+        let quitMsg = "Are you sure you want to quit?"
+         in ( ""
+            , padBottom (Pad 1) $ hCenter $ txt quitMsg
+            , Just (0, [("Cancel", False), ("Quit", True)])
```

Comment:
As written in my review for `handleModalEvent` I would consider replacing `False` and `True` with values `ButtonCancel` and `ButtonQuit` of a `ButtonSelection` type in order to clarify the semantics.

Reply:
I just called them `Cancel` and `Confirm` to try to make it slightly more general in case we later want another type of dialog that asks you to confirm something.